File: cpp/base/src/vsdk/toolkit/render/RadianceImageExporter.cpp
#include "vsdk/toolkit/common/logging/Logger.h"
#include "vsdk/toolkit/common/linealAlgebra/Numeric.h"
#include "vsdk/toolkit/io/image/ImageOutputHandle.h"
#include "vsdk/toolkit/scene/RadianceMethod.h"
#include "vsdk/toolkit/render/RadianceImageExporter.h"
#include "vsdk/toolkit/render/ScreenBuffer.h"
#include "vsdk/toolkit/render/SoftIdsWrapper.h"

void
RadianceImageExporter::exportImage(
    const char *fileName,
    java::OutputStream *outputStream,
    const int isPipe,
    const Scene *scene,
    const RadianceMethod *radianceMethod,
    ToneMappingContext *toneMapOptions,
    const RendererConfiguration *renderOptions)
{
    if ( outputStream == nullptr || scene == nullptr || scene->camera == nullptr ) {
        return;
    }

    if ( toneMapOptions == nullptr ) {
        Logger::error("RadianceImageExporter::exportImage", "Tone mapping context not provided for image export");
        return;
    }

    ScreenBuffer screenBuffer(nullptr, scene->camera, toneMapOptions);
    SoftIdsWrapper idRenderer(scene, renderOptions);

    long width;
    long height;
    idRenderer.getSize(&width, &height);
    if ( width != screenBuffer.getHRes() || height != screenBuffer.getVRes() ) {
        Logger::error("RadianceImageExporter::exportImage", "ID buffer size does not match screen size");
        return;
    }

    for ( int y = 0; y < height; y++ ) {
        for ( int x = 0; x < width; x++ ) {
            Patch *patch = idRenderer.getPatchAtPixel(x, y);
            if ( patch != nullptr ) {
                const ColorRgb radiance = RadianceImageExporter::getRadianceAtPixel(
                    &screenBuffer,
                    scene->camera,
                    x,
                    y,
                    patch,
                    radianceMethod,
                    renderOptions);
                screenBuffer.add(x, y, radiance);
            }
        }
    }

    ImageOutputHandle *imageOutputHandle = ImageOutputHandle::createRadianceImageOutputHandle(
        fileName,
        outputStream,
        isPipe,
        scene->camera->xSize,
        scene->camera->ySize);
    if ( imageOutputHandle == nullptr ) {
        return;
    }

    screenBuffer.writeFile(imageOutputHandle);
    ImageOutputHandle::deleteImageOutputHandle(imageOutputHandle);
}

void
RadianceImageExporter::clipUv(int numberOfVertices, double *u, double *v) {
    if ( *u > 1.0 - Numeric::EPSILON ) {
        *u = 1.0 - Numeric::EPSILON;
    }
    if ( *v > 1.0 - Numeric::EPSILON ) {
        *v = 1.0 - Numeric::EPSILON;
    }
    if ( numberOfVertices == 3 && (*u + *v) > 1.0 - Numeric::EPSILON ) {
        if ( *u > *v ) {
            *u = 1.0 - *v - Numeric::EPSILON;
        } else {
            *v = 1.0 - *u - Numeric::EPSILON;
        }
    }
    if ( *u < Numeric::EPSILON ) {
        *u = Numeric::EPSILON;
    }
    if ( *v < Numeric::EPSILON ) {
        *v = Numeric::EPSILON;
    }
}

ColorRgb
RadianceImageExporter::getRadianceAtPixel(
    const ScreenBuffer *screenBuffer,
    Camera *camera,
    int x,
    int y,
    Patch *patch,
    const RadianceMethod *radianceMethod,
    const RendererConfiguration *renderOptions)
{
    ColorRgb radiance{};
    radiance.clear();

    if ( screenBuffer == nullptr || camera == nullptr || patch == nullptr || radianceMethod == nullptr || renderOptions == nullptr ) {
        return radiance;
    }

    Vector3D rayDirection = screenBuffer->getPixelVector(x, y);
    rayDirection.normalize(Numeric::EPSILON_FLOAT);

    const float denominator = patch->getNormal().dotProduct(rayDirection);
    if ( denominator <= Numeric::EPSILON_FLOAT && denominator >= -Numeric::EPSILON_FLOAT ) {
        return radiance;
    }

    const float distance =
        -(patch->getNormal().dotProduct(camera->eyePosition) + patch->getPlaneConstant()) / denominator;
    Vector3D hitPoint;
    hitPoint.sumScaled(camera->eyePosition, distance, rayDirection);

    double u;
    double v;
    patch->uv(&hitPoint, &u, &v);
    RadianceImageExporter::clipUv(patch->getNumberOfVertices(), &u, &v);

    const Vector3D eyeDirection(-rayDirection.x, -rayDirection.y, -rayDirection.z);
    return radianceMethod->getRadiance(camera, patch, u, v, eyeDirection, renderOptions);
}

File: typescript/base/src/main/vsdk/toolkit/render/RadianceImageExporter.ts
import { OutputStream } from "../../../java/io/OutputStream";
import { ColorRgb } from "../common/color/ColorRgb";
import { Logger } from "../common/logging/Logger";
import { RendererConfiguration } from "../material/RendererConfiguration";
import { Numeric } from "../common/linealAlgebra/Numeric";
import { Vector3D } from "../common/linealAlgebra/Vector3D";
import { ImageOutputHandle } from "../io/image/ImageOutputHandle";
import { Camera } from "../scene/Camera";
import { RadianceMethod } from "../scene/RadianceMethod";
import { Scene } from "../scene/Scene";
import { Patch } from "../environment/geometry/elements/Patch";
import { ToneMappingContext } from "../tonemap/ToneMappingContext";
import { ScreenBuffer } from "./ScreenBuffer";
import { SoftIdsWrapper } from "./SoftIdsWrapper";

export class RadianceImageExporter {
  private constructor() {
  }

  public static exportImage(
    fileName: string,
    outputStream: OutputStream | null,
    isPipe: number,
    scene: Scene | null,
    radianceMethod: RadianceMethod,
    toneMapOptions: ToneMappingContext | null,
    renderOptions: RendererConfiguration
  ): void {
    if (outputStream === null || scene === null || scene.camera === null) {
      return;
    }

    if (toneMapOptions === null) {
      Logger.error("RadianceImageExporter::exportImage", "Tone mapping context not provided for image export");
      return;
    }

    const screenBuffer = new ScreenBuffer(null, scene.camera, toneMapOptions);
    const idRenderer = new SoftIdsWrapper(scene, renderOptions);

    const width = [0];
    const height = [0];
    idRenderer.getSize(width, height);
    if (width[0] !== screenBuffer.getHRes() || height[0] !== screenBuffer.getVRes()) {
      Logger.error("RadianceImageExporter::exportImage", "ID buffer size does not match screen size");
      return;
    }

    for (let y = 0; y < height[0]; y++) {
      for (let x = 0; x < width[0]; x++) {
        const patch = idRenderer.getPatchAtPixel(x, y);
        if (patch !== null) {
          const radiance = RadianceImageExporter.getRadianceAtPixel(
            screenBuffer,
            scene.camera,
            x,
            y,
            patch,
            radianceMethod,
            renderOptions
          );
          screenBuffer.add(x, y, radiance);
        }
      }
    }

    const imageOutputHandle = ImageOutputHandle.createRadianceImageOutputHandle(
      fileName,
      outputStream,
      isPipe,
      scene.camera.xSize,
      scene.camera.ySize
    );
    if (imageOutputHandle === null) {
      return;
    }

    screenBuffer.writeFile(imageOutputHandle);
    ImageOutputHandle.deleteImageOutputHandle(imageOutputHandle);
  }

  private static clipUv(numberOfVertices: number, u: number[], v: number[]): void {
    if (u[0] > 1.0 - Numeric.EPSILON) {
      u[0] = 1.0 - Numeric.EPSILON;
    }
    if (v[0] > 1.0 - Numeric.EPSILON) {
      v[0] = 1.0 - Numeric.EPSILON;
    }
    if (numberOfVertices === 3 && (u[0] + v[0]) > 1.0 - Numeric.EPSILON) {
      if (u[0] > v[0]) {
        u[0] = 1.0 - v[0] - Numeric.EPSILON;
      }
      else {
        v[0] = 1.0 - u[0] - Numeric.EPSILON;
      }
    }
    if (u[0] < Numeric.EPSILON) {
      u[0] = Numeric.EPSILON;
    }
    if (v[0] < Numeric.EPSILON) {
      v[0] = Numeric.EPSILON;
    }
  }

  private static getRadianceAtPixel(
    screenBuffer: ScreenBuffer | null,
    camera: Camera | null,
    x: number,
    y: number,
    patch: Patch | null,
    radianceMethod: RadianceMethod | null,
    renderOptions: RendererConfiguration | null
  ): ColorRgb {
    const radiance = new ColorRgb();
    radiance.clear();

    if (screenBuffer === null || camera === null || patch === null || radianceMethod === null || renderOptions === null) {
      return radiance;
    }

    const rayDirection = screenBuffer.getPixelVector(x, y);
    rayDirection.normalize(Numeric.EPSILON_FLOAT);

    const denominator = patch.normal.dotProduct(rayDirection);
    if (denominator <= Numeric.EPSILON_FLOAT && denominator >= -Numeric.EPSILON_FLOAT) {
      return radiance;
    }

    const distance = -(patch.normal.dotProduct(camera.eyePosition) + patch.planeConstant) / denominator;
    const hitPoint = new Vector3D();
    hitPoint.sumScaled(camera.eyePosition, distance, rayDirection);

    const u = [0.0];
    const v = [0.0];
    patch.uv(hitPoint, u, v);
    RadianceImageExporter.clipUv(patch.numberOfVertices, u, v);

    const eyeDirection = new Vector3D(-rayDirection.x, -rayDirection.y, -rayDirection.z);
    return radianceMethod.getRadiance(camera, patch, u[0], v[0], eyeDirection, renderOptions);
  }
}
